### PR TITLE
gazebo_plugins: Port the gazebo_ros_p3d plugin

### PR DESCRIPTION
**Pull request by mjeronimo**
This is a port of the gazebo_ros_p3d plugin to ROS2. Tested with gazebo 9.0.0.

Example plugin XML:

```
<plugin name="p3d_base_controller" filename="libgazebo_ros_p3d.so">
  <ros>
    <namespace></namespace>
    <argument></argument>
  </ros>
  <update_rate>50.0</update_rate>
  <body_name>base_link</body_name>
  <topic_name>base_pose_ground_truth</topic_name>
  <frame_name>world</frame_name>
  <xyz_offsets>0 0 0</xyz_offsets>
  <rpy_offsets>0 0 0</rpy_offsets>
  <gaussian_noise>0.01</gaussian_noise>
</plugin>
```

### CODE REVIEW

File: gazebo_plugins/include/gazebo_plugins/gazebo_ros_p3d.hpp
Comment (chapulina) [inline gazebo_plugins/include/gazebo_plugins/gazebo_ros_p3d.hpp]: Includes from external libraries should use `<>`

File: gazebo_plugins/src/gazebo_ros_p3d.cpp
Comment (chapulina) [inline gazebo_plugins/src/gazebo_ros_p3d.cpp]: We're completing documentation as we port plugins to ROS 2, mind documenting all members?

File: gazebo_plugins/src/gazebo_ros_p3d.cpp
Comment (chapulina) [inline gazebo_plugins/src/gazebo_ros_p3d.cpp]: I was going to say to use `std::mutex`, but actually I think this mutex can be completely removed. It is only used within the Update callback which always happens in the same thread (gazebo's physics thread), and unlike what the comment says, there are no message callbacks in this plugin.

File: gazebo_plugins/src/gazebo_ros_p3d.cpp
Comment (chapulina) [inline gazebo_plugins/src/gazebo_ros_p3d.cpp]: It looks like the seed is always zero and there's no way to set it. I suggest this variable be removed.

File: gazebo_plugins/src/gazebo_ros_p3d.cpp
Comment (chapulina) [inline gazebo_plugins/src/gazebo_ros_p3d.cpp]: This line shouldn't be necessary

File: gazebo_plugins/src/gazebo_ros_p3d.cpp
Comment (chapulina) [inline gazebo_plugins/src/gazebo_ros_p3d.cpp]: `model_` is only used in this function, we don't need to keep a shared pointer to it

File: gazebo_plugins/src/gazebo_ros_p3d.cpp
Comment (chapulina) [inline gazebo_plugins/src/gazebo_ros_p3d.cpp]: ROS 2 wil only support Gazebo 9+, so you can get rid of all the if defs

File: gazebo_plugins/src/gazebo_ros_p3d.cpp
Comment (chapulina) [inline gazebo_plugins/src/gazebo_ros_p3d.cpp]: You can change this callback to take `gazebo::common::UpdateInfo`, so you'll get the sim time as an input parameter. This also means there's no more need to keep the `world_` pointer. [Here](https://github.com/ros-simulation/gazebo_ros_pkgs/blob/ros2/gazebo_plugins/src/gazebo_ros_joint_state_publisher.cpp#L140)'s an example.

File: gazebo_plugins/src/gazebo_ros_p3d.cpp
Comment (chapulina) [inline gazebo_plugins/src/gazebo_ros_p3d.cpp]: How about returning here if `impl_->ros_node_->count_subscribers(impl_->topic_name_) == 0`? This should make the code a bit more readable.

Comment (chapulina) [reply gazebo_plugins/src/gazebo_ros_p3d.cpp]: The same goes for `tmp_dt == 0`, doing this would reduce 2 levels of indentation

File: gazebo_plugins/src/gazebo_ros_p3d.cpp
Comment (chapulina) [inline gazebo_plugins/src/gazebo_ros_p3d.cpp]: Yeah it looks like `apos_`, `aeul_`, `frame_apos_` and `frame_aeul_` can be completely removed.

File: gazebo_plugins/src/gazebo_ros_p3d.cpp
Comment (chapulina) [inline gazebo_plugins/src/gazebo_ros_p3d.cpp]: Is there a reason to keep this as a member variable? Can't a new message be created at every update? It doesn't look like we need past values or anything like that.

**Review by chapulina — CHANGES_REQUESTED** [review]
Thank you again for the PR, @mjeronimo ! I know I left a lot of comments, but that's mainly our fault, not yours. A lot has changed from ROS 1 to ROS 2, and although we've been writing migration guides for downstream developers, we're lacking a guide specific for contributors like you. I'm going to work on this soon.

Meanwhile, besides the comments I left:

- [x] Please remove the files from `.ros1_unported` which you've ported
- [x] Please add a basic test for the plugin. You should be able to start from one of the [tests for plugins already ported](https://github.com/ros-simulation/gazebo_ros_pkgs/tree/ros2/gazebo_plugins/test).
- [x] Also add a demo world with instructions similar to the ones [here](https://github.com/ros-simulation/gazebo_ros_pkgs/tree/ros2/gazebo_plugins/worlds).
- [x] Add usage documentation to the header, see the [other plugins](https://github.com/ros-simulation/gazebo_ros_pkgs/tree/ros2/gazebo_plugins/include/gazebo_plugins) for examples.
- [x] Add a migration guide for downstream users to the [wiki](https://github.com/ros-simulation/gazebo_ros_pkgs/wiki).

I realize this seems like a lot, but hopefully the existing plugins will take you a long way. Don't hesitate to ask questions.

File: gazebo_plugins/src/gazebo_ros_p3d.cpp
Comment (mjeronimo) [reply gazebo_plugins/src/gazebo_ros_p3d.cpp]: Removed.

File: gazebo_plugins/src/gazebo_ros_p3d.cpp
Comment (mjeronimo) [reply gazebo_plugins/src/gazebo_ros_p3d.cpp]: Removed.

File: gazebo_plugins/src/gazebo_ros_p3d.cpp
Comment (mjeronimo) [reply gazebo_plugins/src/gazebo_ros_p3d.cpp]: Removed.

File: gazebo_plugins/src/gazebo_ros_p3d.cpp
Comment (mjeronimo) [reply gazebo_plugins/src/gazebo_ros_p3d.cpp]: Removed.

File: gazebo_plugins/src/gazebo_ros_p3d.cpp
Comment (mjeronimo) [reply gazebo_plugins/src/gazebo_ros_p3d.cpp]: Updated to not be a member variable.

File: gazebo_plugins/include/gazebo_plugins/gazebo_ros_p3d.hpp
Comment (mjeronimo) [reply gazebo_plugins/include/gazebo_plugins/gazebo_ros_p3d.hpp]: Sure, updated to follow that convention.

File: gazebo_plugins/src/gazebo_ros_p3d.cpp
Comment (mjeronimo) [reply gazebo_plugins/src/gazebo_ros_p3d.cpp]: Good idea. Previously, I was concerned about a straight port and not improvements. Done.

File: gazebo_plugins/src/gazebo_ros_p3d.cpp
Comment (mjeronimo) [reply gazebo_plugins/src/gazebo_ros_p3d.cpp]: Removed.

File: gazebo_plugins/src/gazebo_ros_p3d.cpp
Comment (mjeronimo) [reply gazebo_plugins/src/gazebo_ros_p3d.cpp]: Removed altogether since it was only an alias for _parent.

File: gazebo_plugins/src/gazebo_ros_p3d.cpp
Comment (mjeronimo) [reply gazebo_plugins/src/gazebo_ros_p3d.cpp]: I've removed these new members with other changes, so this is no longer applicable.

File: gazebo_plugins/src/gazebo_ros_p3d.cpp
Comment (mjeronimo) [reply gazebo_plugins/src/gazebo_ros_p3d.cpp]: Done; world_ pointer removed.

**Comment by chapulina** [comment]
@mjeronimo , I went ahead and tackled the remaining items in the list:

* Rebased with latest `ros2`
* Added a test in https://github.com/ros-simulation/gazebo_ros_pkgs/pull/845/commits/662081ab921601525535bb44630732ecc339a95b
* Added a demo world in https://github.com/ros-simulation/gazebo_ros_pkgs/pull/845/commits/662081ab921601525535bb44630732ecc339a95b
* Added usage documentation in https://github.com/ros-simulation/gazebo_ros_pkgs/pull/845/commits/662081ab921601525535bb44630732ecc339a95b
* Added a migration guide: [ROS 2 Migration: P3D](https://github.com/ros-simulation/gazebo_ros_pkgs/wiki/ROS-2-Migration:-P3D)

Let me know if you have any concerns, I'm ready to merge once you approve the latest changes.

By the way, I also added a contribution guide [here](https://github.com/ros-simulation/gazebo_ros_pkgs/wiki/ROS-2-migration-contribution-guide).

**Review by chapulina — APPROVED** [review]

**Comment by mjeronimo** [comment]
@chapulina  Thanks! Looks good to me. I was planning to get back to this after vacation (just back today), but glad to see you've taken care of it.

**Pull request merged**
